### PR TITLE
puller: fix a memory leak bug in the frontier

### LINKED ISSUE #683: There is a memory leak in the frontier
## Bug Report

Please answer these questions before submitting your issue. Thanks!

In the current implementation(master, v4.0.2), after region splitting, we add some related nodes to the frontier; 
however, after region merging, we do not remove the overwritten nodes, which leads to an increasing number of nodes maintained in the frontier.

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 
fix a memory leak bug in the frontier

after this pr, nodes in the frontier could only be added, not removed, which led to the memory leak issue

In this pr, when regions are merged, the frontier will remove redundant nodes and keeps only the necessary nodes

fix #683

#### Before this PR
```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/ticdc/cdc/puller/frontier
BenchmarkSpanFrontier/5k-8           6533013           183 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontier/10k-8          5940601           206 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontier/50k-8          4927638           252 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontier/100k-8         4529256           242 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/5k_5-8              5396868           213 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/5k_10-8             4033922           300 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/5k_100-8            1000000          1169 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/5k_500-8             251289          4662 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/10k_5-8             4996110           230 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/10k_10-8            4038318           309 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/10k_100-8           1000000          1123 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/10k_500-8            251916          5005 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/50k_5-8             4139940           319 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/50k_10-8            3498513           364 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/50k_100-8           1000000          1143 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/50k_500-8            248053          4866 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/100k_5-8            3539360           316 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/100k_10-8           3341758           366 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/100k_100-8          1000000          1225 ns/op           0 B/op          0 allocs/op
BenchmarkSpanFrontierOverlap/100k_500-8           248091          5223 ns/op           0 B/op          0 allocs/op
```

#### After This PR

```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/ticdc/cdc/puller/frontier
BenchmarkSpanFrontier/5k-8       	 4541343	       281 ns/op	      96 B/op	       1 allocs/op
BenchmarkSpanFrontier/10k-8      	 4640968	       261 ns/op	      96 B/op	       1 allocs/op
BenchmarkSpanFrontier/50k-8      	 3709297	       319 ns/op	      96 B/op	       1 allocs/op
BenchmarkSpanFrontier/100k-8     	 3390427	       337 ns/op	      96 B/op	       1 allocs/op
BenchmarkSpanFrontierOverlap/5k_5-8         	 1335870	       796 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/5k_10-8        	 1721634	       697 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/5k_100-8       	 1721421	       698 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/5k_500-8       	 1768676	       684 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/10k_5-8        	 1570315	       773 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/10k_10-8       	 1536466	       800 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/10k_100-8      	 1515050	       838 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/10k_500-8      	 1584950	       797 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/50k_5-8        	 1426015	       877 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/50k_10-8       	 1377284	       903 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/50k_100-8      	 1381473	       834 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/50k_500-8      	 1408707	       873 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/100k_5-8       	 1346317	       913 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/100k_10-8      	 1337869	       896 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/100k_100-8     	 1356310	       948 ns/op	     374 B/op	       6 allocs/op
BenchmarkSpanFrontierOverlap/100k_500-8     	 1435084	       841 ns/op	     374 B/op	       6 allocs/op
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: cdc/puller/frontier/frontier_bench_test.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func BenchmarkSpanFrontier(b *testing.B) {
 				spans = append(spans, span)
 			}
 
-			f := NewFrontier(spans...)
+			f := NewFrontier(0, spans...)
```
Comment (amyangfei) [inline cdc/puller/frontier/frontier_bench_test.go:52]: Do we have a benchmark result after this PR.

File: cdc/puller/frontier/list.go [hunk 2]
```diff
@@ -24,47 +27,60 @@ const (
...
+func newSpanList() *skipList {
+	l := new(skipList)
+	l.head.nexts = make([]*skipListNode, maxHeight)
+	return l
+}
+
+func (l *skipList) randomHeight() int {
 	h := 1
-	for h < maxHeight && fastrand() < uint32(math.MaxUint32)/4 {
+	for h < maxHeight && fastrand() < uint32(math.MaxUint32)/2 {
```
Comment (amyangfei) [inline cdc/puller/frontier/list.go:83]: What are the effects of this change?

Comment (leoppro) [reply cdc/puller/frontier/list.go:83]: it’s for debugging, I will recover it

File: cdc/puller/frontier/frontier_bench_test.go (hunk 1 above)
Comment (leoppro) [reply cdc/puller/frontier/frontier_bench_test.go:52]: updated at this issue

**Pull request merged**